httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheControlHeaderParser.java

```diff
@@ -0,0 +1,205 @@
+
+    /**
+     * The logger for this class.
+     */
+    private static final Logger LOG = LoggerFactory.getLogger(CacheControlHeaderParser.class);
+
+    /**
+     * The delimiter character used to separate caching directives in the header.
+     */
+    private static final char PARAM_DELIMITER = ';';
+
+    /**
+     * The delimiter character used to separate values within a caching directive.
+     */
+    private static final char VALUE_DELIMITER = '=';
+
+    /**
+     * The set of characters that can delimit a token in the header.
+     */
+    private static final BitSet TOKEN_DELIMS = Tokenizer.INIT_BITSET(VALUE_DELIMITER, PARAM_DELIMITER, ',');
+
+    /**
+     * The set of characters that can delimit a value in the header.
+     */
+    private static final BitSet VALUE_DELIMS = Tokenizer.INIT_BITSET(VALUE_DELIMITER, PARAM_DELIMITER, ',');
+
+    /**
+     * The token parser used to extract values from the header.
+     */
+    private final Tokenizer tokenParser;
+
+    /**
+     * Constructs a new instance of this parser.
+     */
+    protected CacheControlHeaderParser() {
+        super();
+        this.tokenParser = Tokenizer.INSTANCE;
+    }
+
+    /**
+     * Parses the specified header and returns a {@link CacheControlHeader} object containing the relevant caching
+     * directives.
+     *
+     * @param header the header to parse
+     * @return a {@link CacheControlHeader} object containing the relevant caching directives
+     * @throws IllegalArgumentException if the header is {@code null}
+     * @throws MalformedCookieException if the value of the "max-age" or "s-maxage" field is malformed
+     */
+    public final CacheControlHeader parse(final Header header) throws MalformedCookieException {
+       Args.notNull(header, "Header");
+
+        final CacheControlHeader cacheControlHeader = new CacheControlHeader();
+        final CharArrayBuffer buffer;
+        final Tokenizer.Cursor cursor;
+        if (header instanceof FormattedHeader) {
+            buffer = ((FormattedHeader) header).getBuffer();
+            cursor = new Tokenizer.Cursor(((FormattedHeader) header).getValuePos(), buffer.length());
+        } else {
+            final String s = header.getValue();
+            if (s == null) {
+                return cacheControlHeader;
+            }
+            buffer = new CharArrayBuffer(s.length());
+            buffer.append(s);
+            cursor = new Tokenizer.Cursor(0, buffer.length());
+        }
+
+        while (!cursor.atEnd()) {
+            final String name = tokenParser.parseToken(buffer, cursor, TOKEN_DELIMS);
+            if (name.isEmpty()) {
+                continue;
+            }
+            if (cursor.atEnd()) {
+                return cacheControlHeader;
+            }
+            final int valueDelim = buffer.charAt(cursor.getPos());
+            cursor.updatePos(cursor.getPos() + 1);
+            if (valueDelim != EQUAL_CHAR) {
+                if (LOG.isWarnEnabled()) {
```

Comment:
@arturobernalg Do we really need this warning? Is the header really invalid or is it a directive without a parameter?